[clang] Fix implicit integer conversion for opaque enums declared in class templates

thebrandre opened the pull request:
This commit fixes issues with enumeration types instantiated from an opaque-enum-declarations 
(see [dcl.enum]) in class templates broke basic assumptions during parsing of arithmetic
expressions due to absent (NULL TYPE) promotion types of instances of EnumDecl.

To this end, we repeat the simple steps in `Sema::ActOnTag` to evaluate the promotion type
of a fixed enumeration based on its underlying type (see C++11 [conv.prom] p4).

Note that if, instead, a full *enum-specifier* (subsequent curly braces) is provided,
`Sema::ActOnEnumBody` is re-invoked on template instantiation anyway overriding the
promotion type and hiding the issue. This is analog to how enumerations declarations
outside of template declarations are handled.
Note that, in contrast to `Sema::ActOnEnumBody`, `Sema::ActOnTag` is *not* called again
for the instantiated enumeration type.

Fixes #117960.

cor3ntin commented:
Thanks for the PR. Can you update the commit message to be more descriptive?

I'm not sure the fix is sufficient.
I think most of the checks done in `ActOnEnumBody` should be (re) performed on instantiation.

Maybe we should instead

- Add a new `BuildEnumBody` function, move most (all?) the implementation of  `ActOnEnumBody` to that
- call `BuildEnumBody` from both `ActOnEnumBody` and `RebuildEnumType`

That would not only fix the crash but also the fact that we are seemingly missing a lot of diagnostics
https://godbolt.org/z/17dTW4dEe

thebrandre commented:
Thanks for the feedback @cor3ntin! 🙂

I will change the title of the PR and of the commit message to: 

> _[clang] Fix implicit integer conversion for opaque enum declarations in class templates_ 

adopting the terminology from [[dcl.enum]](https://wg21.link/dcl.enum) if that isn't too verbose?

The other potential issues you mentioned should actually conform to the specification in [[temp.spec.general]](https://wg21.link/temp.spec#temp.inst-3) if I understand it correctly.
See here on Compiler Explorer: [https://godbolt.org/z/GzYq9jzjo](https://godbolt.org/z/GzYq9jzjo).
This behavior is also consistent with GCC.

My changes should not affect the behavior if `ActOnEnumBody` is called. Admittedly, I was very conservative when changing behavior because I am not yet very familiar with the codebase.
Do you think, we should still pursue the idea of refactoring as you described if the current behavior of Clang turns out to be correct in the situations you brought up? I'd be happy to do so. But I am not sure if it belongs into this PR? It definitely makes sense to add regression tests for this scenario if they don't already exist.

In any case, I'll try if I can find any other related scenarios where the fix isn't sufficient!

thebrandre commented:
I added regression tests for the scoped enumeration types to make sure that these are not affected.
And I also tried to improve the commit message and added comments with references to the relevant sections in the C++ standard.
But I still have to think about your refactoring/completeness suggestions.

thebrandre committed:
[clang] Fix missing promotion type for opaque enum declarations in class templates

Repeat the steps in `Sema::ActOnTag` for the promotion type after type substitution of the
underlying type. This is required in the case of an *opaque-enum-declaration* (see [dcl.enum]).
If, instead, a full *enum-specifier* (subsequent curly braces) is provided, `Sema::ActOnEnumBody`
is re-invoked on template instantiation overriding the promotion type and hiding the issue.
Note that, in contrast to `Sema::ActOnEnumBody`, `Sema::ActOnTag` is *not* called again for the
instantiated enumeration type.

Fixes #117960.

thebrandre committed:
[clang] Add tests for implicit integer conversion of enums declared in class templates

These tests are variants of the crash reported in GitHub issue #117960,
in which an opaque-enum-declaration in a class template broke basic assumptions
during parsing of arithmetic expressions due to missing promotion types
of instances of EnumDecl.

thebrandre commented:
@cor3ntin Good news! I figured it out on my own! And I now have a fix that I am confident in. 🙂
I analyzed the steps of the parser in **a lot**  of similar cases, for which I also added quite a few regression tests.

The attributes have already been handled correctly previously. But to make sure of it, I also added tests.

Putting the few lines for the fix there, seems fairly consistent to me. It repeats the logic in [`Sema::ActOnTag`](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.6/clang/lib/Sema/SemaDecl.cpp#L17152-L17154) and [`Sema::ActOnEnumBody`](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.6/clang/lib/Sema/SemaDecl.cpp#L19917-L19922) (exact lines of code are linked). We would now repeat them a third time but this is unavoidable because `Sema::ActOnEnumBody` is called again during instantiation but `Sema::ActOnTag` is not.
Refactoring this is would be hard because it's not really specified what exactly the individual functions are supposed to do. And the repetition doesn't seem too bad to me.

As you suggested, I stepped through the code to check if anything else from `Sema::ActOnTag` might be missing. But this doesn't seem to be the case.
Some things seemed odd but non-relevant. For example for `enum E : int {};`  `EnumDecl::NumPositiveBits` is set to 1, but it stays 0 for the opaque declaration `enum E : int;`. I also noticed that I can trigger a case where [a diagnostic gets reported twice](https://godbolt.org/z/473Eezhjn)). This, however, seems rather unrelated to me. I would rather open up a separate pull request for that.

Thanks again for your feedback @cor3ntin - it helped me to get a much cleaner solution and a better understanding of the issue even though my initial analysis might have been a little confusing ...
Well, I hope it's now ready for merging. 🙂

I rebased my branch on the current main and also "rebased away" my prior fix to make the diff more concise.

cor3ntin commented:
> Thanks for the PR. Can you update the commit message to be more descriptive?
> 
> I'm not sure the fix is sufficient.
> I think most of the checks done in `ActOnEnumBody` should be (re) performed on instantiation.
> 
> Maybe we should instead
> 
> - Add a new `BuildEnumBody` function, move most (all?) the implementation of  `ActOnEnumBody` to that
> - call `BuildEnumBody` from both `ActOnEnumBody` and `RebuildEnumType`
> 
> That would not only fix the crash but also the fact that we are seemingly missing a lot of diagnostics
> https://godbolt.org/z/17dTW4dEe
> 

@erichkeane wdyt?

thebrandre commented:
@cor3ntin I hope there is no misunderstanding here. If so, I apologize,

I strongly agree with you that some refactoring should be applied because, as I mentioned, the diagnostics get duplicated in certain cases as shown here on Compiler Explorer: [https://godbolt.org/z/473Eezhjn](https://godbolt.org/z/473Eezhjn).
But I would prefer opening another pull request for it keeping it separate from the fix making it easier to cherry-pick.
My understanding is that all commits get squashed before being applied to main?

If there is still something wrong with the commit message, I would gladly give it another try if you could be more specific.

erichkeane commented:
> > Thanks for the PR. Can you update the commit message to be more descriptive?
> > I'm not sure the fix is sufficient.
> > I think most of the checks done in `ActOnEnumBody` should be (re) performed on instantiation.
> > Maybe we should instead
> > 
> > * Add a new `BuildEnumBody` function, move most (all?) the implementation of  `ActOnEnumBody` to that
> > * call `BuildEnumBody` from both `ActOnEnumBody` and `RebuildEnumType`
> > 
> > That would not only fix the crash but also the fact that we are seemingly missing a lot of diagnostics
> > https://godbolt.org/z/17dTW4dEe
> 
> @erichkeane wdyt?

I agree, it seems that the instantiation of the enum is missing quite a bit when it comes to diagnostics.  We should definitely do some sort of analysis to ensure that we're getting the same diagnostics at both points.

That said, I am ok with doing this NOW, as long as the author promises to do the rest in a followup patch ASAP.

thebrandre commented:
@erichkeane I am not sure what you mean. As I pointed out earlier, the examples @cor3ntin brought up actually conform to the standard as in [[temp.spec.general]](https://wg21.link/temp.spec#temp.inst-3):

> The implicit instantiation of a class template specialization causes the implicit instantiation of the declarations, but not of the definitions of [...] scoped member enumerations, [...]

The issues are correctly diagnosed by clang if the struct is *explicitly instantiated* as demonstrated here in Compiler Explorer: [https://godbolt.org/z/GzYq9jzjo](https://godbolt.org/z/GzYq9jzjo).

Nevertheless, I investigated further and managed to discover some real bugs with diagnostics on template instantiation.
These affect not only enums but also nested classes. If you are interested, you can have a look: [https://godbolt.org/z/vKG6vqWhs](https://godbolt.org/z/vKG6vqWhs). I have already started working on these issues but so far they seem unrelated.

I plan to keep working on it but I can't promise much given the facts that I only have the weekends to spend time on it and my limited familiarity with the codebase (this would be my very first commit to LLVM). But, of course, I promise to try my best! 🙂

erichkeane commented:
> @erichkeane I am not sure what you mean. As I pointed out earlier, the examples @cor3ntin brought up actually conform to the standard as in [[temp.spec.general]](https://wg21.link/temp.spec#temp.inst-3):
> 
> > The implicit instantiation of a class template specialization causes the implicit instantiation of the declarations, but not of the definitions of [...] scoped member enumerations, [...]
> 
> The issues are correctly diagnosed by clang if the struct is _explicitly instantiated_ as demonstrated here in Compiler Explorer: https://godbolt.org/z/GzYq9jzjo.

My quick look through the ActOnEnum and this function looked like it showed some misses.  If Corentin is comfortable with our checking here, then I'm happy too.

> 
> Nevertheless, I investigated further and managed to discover some real bugs with diagnostics on template instantiation. These affect not only enums but also nested classes. If you are interested, you can have a look: https://godbolt.org/z/vKG6vqWhs. I have already started working on these issues but so far they seem unrelated.

Thats great, thanks for finding that and working on it!  Those look like some pretty unfortunate misses.
> 
> I plan to keep working on it but I can't promise much given the facts that I only have the weekends to spend time on it and my limited familiarity with the codebase (this would be my very first commit to LLVM). But, of course, I promise to try my best! 🙂

Definitely appreciate it!

I'm happy when Corentin is here, I have no concerns.

cor3ntin commented:
We usually try to diagnose issues as soon as possible, so i still believe there are improvements to be made here. That being said, I'm happy to approve the PR as it is now. 

That being said, it is missing a changelog entry in `clang/docs/ReleaseNotes.rst`, can you add one?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -1571,6 +1571,23 @@ Decl *TemplateDeclInstantiator::VisitEnumDecl(EnumDecl *D) {
         Enum->setIntegerType(SemaRef.Context.IntTy);
       else
         Enum->setIntegerTypeSourceInfo(NewTI);
+
+      // The following lines are relevant for opaque-enum-declarations.
+      // If users declare a full enum-specifier, the promotion type is reset
+      // again when parsing the (potentially empty) enumerator-list.
+      // We implement the same logic from C++11 [conv.prom] p4 here but only
+      // after template specialization [temp.spec.general] because it requires
+      // the concrete type.
+      // Note that (1) this also correctly handles the non-dependent case,
+      // (2) we set the promotion type for scoped enumerations to ensure
+      // consistency with declarations outside of templates, (3) we guarantee a
+      // valid promotion even if the user provided an invalid underlying type
+      // (fallback to "default int").
```
```suggestion
      // C++23 [conv.prom]p4
      // if integral promotion can be applied to its underlying type, a prvalue of an unscoped enumeration type 
      // whose underlying type is fixed can also be converted to a prvalue of the promoted underlying type.
      // 
      // FIXME: that logic is already implemented in ActOnEnumBody, factor out into (Re)BuildEnumBody. 
```

cor3ntin commented:
The issue number here seems wrong, can you update the description?
https://github.com/llvm/llvm-project/pull/121039#issue-2757453924

thebrandre committed:
[clang] Add release notes for resolving issue #117960

thebrandre committed:
[clang] Reduce verbosity of comment and add FIXME

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

thebrandre committed:
[clang] Reformat code

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -938,6 +938,9 @@ Miscellaneous Clang Crashes Fixed
   module is found through ADL. (GH#109879)
 
+- Fixed a crash when an unscoped enumeration declared by an opaque-enum-declaration within a class template
+  with a dependent underlying type is subject to integral promotion. (GH#117960)
```
```suggestion
  with a dependent underlying type is subject to integral promotion. (#GH117960)
```

thebrandre commented:
> The issue number here seems wrong, can you update the description? [#121039 (comment)](https://github.com/llvm/llvm-project/pull/121039#issue-2757453924) (#116525 is a completely unrelated issue)

Thanks for pointing that out @cor3ntin. I messed that up in my last edit.

> That being said, it is missing a changelog entry in clang/docs/ReleaseNotes.rst, can you add one?

Sure. You won't notice it but I literally spent half an hour thinking about this sentence. It's so difficult to be clear, technically correct, and concise at the same time.

> We usually try to diagnose issues as soon as possible, so i still believe there are improvements to be made here. 

I am still undecided here. As far as I understand the standard, this is valid code as long as the enum is not used. It says *The implicit instantiation of a class template specialization causes [...] the implicit instantiation of the definitions of unscoped member enumerations* [[temp.inst]](https://eel.is/c++draft/temp.inst). This feature extensively used for member functions. For example std::vector::resize is not valid for types that are not default-constructible but you can still use std::vector as long as you don't use the bad member functions. 
See here on Compiler Explorer [https://godbolt.org/z/Mn3Y1Ka14](https://godbolt.org/z/Mn3Y1Ka14).

Admittedly, I can't think of any sane use case for this feature on unscoped member enumerations ... but there are very creative people out there. 😉

thebrandre committed:
Fix reference to issue in clang/docs/ReleaseNotes.rst

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

thebrandre commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -938,6 +938,9 @@ Miscellaneous Clang Crashes Fixed
 - Fixed internal assertion firing when a declaration in the implicit global
   module is found through ADL. (GH#109879)
```
Oh thanks! I adopted the pattern used in the note before and didn't notice that it doesn't match the others.

thebrandre reviewed (commented):
I am not sure why the jumbled letters in "GH#117960"  are still there and why I can't apply it (it says the change request is outdated). So I am going to fix that manually.

cor3ntin reviewed (approved):
Thanks!
Do you need me to merge that for you?

thebrandre reviewed (commented):
And it's me again. I did review it before but didn't know I was supposed to explicitly mark each file. 🙈

thebrandre commented:
> Thanks! Do you need me to merge that for you?

That would be great! 😊
